File: sdk-python/dex/state.py
# Legacy Materials in this file remain under their original licenses.
# See LEGACY_NOTICES.md.

# Modifications Copyright (c) 2026 Super Durable, Inc.
#
# Modifications after the Legacy Cutoff are licensed under the
# Super Durable Source License 1.0.
# Legacy Materials remain under their original licenses.
# See LICENSE and LEGACY_NOTICES.md.

from __future__ import annotations

from dataclasses import dataclass
from datetime import timedelta
from enum import Enum
from typing import Any, Generic, Protocol, Sequence, TypeVar, cast

from dex._contract_utils import require_name, validate_condition_id

ValueT = TypeVar("ValueT")


class IndexType(Enum):
    KEYWORD = "keyword"
    FULL_TEXT = "full_text"
    KEYWORD_ARRAY = "keyword_array"
    INT = "int"
    DOUBLE = "double"
    BOOL = "bool"
    DATETIME = "datetime"


@dataclass(frozen=True)
class AttributeIndex:
    type: IndexType
    index_key: str = ""


class Context(Protocol):
    @property
    def flow_id(self) -> str: ...

    @property
    def run_id(self) -> str: ...

    @property
    def step_execution_id(self) -> str: ...

    @property
    def from_step_execution_id(self) -> str: ...

    @property
    def attempt(self) -> int: ...

    def has_timer_fired(self, index: int | None = None) -> bool: ...

    def wait_for_method_failed(self) -> bool: ...

    def set_step_execution_local(self, key: str, value: object) -> None: ...

    def get_step_execution_local(
        self, key: str, value_type: type[ValueT]
    ) -> ValueT | None: ...

    def record_event(self, name: str, value: object) -> None: ...

    def _get_attribute(
        self,
        definition: Attribute[ValueT] | AttributeMap[ValueT],
        instance: str | None,
    ) -> ValueT: ...

    def _set_attribute(
        self,
        definition: Attribute[ValueT] | AttributeMap[ValueT],
        instance: str | None,
        value: ValueT,
    ) -> None: ...

    def _delete_attribute(
        self,
        definition: Attribute[object] | AttributeMap[object],
        instance: str | None,
    ) -> None: ...

    def _publish_channel(
        self,
        definition: Channel[ValueT] | ChannelMap[ValueT],
        instance: str | None,
        value: ValueT,
    ) -> None: ...

    def _channel_size(
        self,
        definition: Channel[object] | ChannelMap[object],
        instance: str | None,
    ) -> int: ...

    def _channel_results(
        self,
        definition: Channel[ValueT] | ChannelMap[ValueT],
        instance: str | None,
    ) -> Sequence[ValueT]: ...


@dataclass(frozen=True)
class Attribute(Generic[ValueT]):
    name: str
    value_type: type[ValueT]
    index: AttributeIndex | None = None

    def __post_init__(self) -> None:
        require_name(self.name)

    def get(self, context: Context) -> ValueT:
        return context._get_attribute(self, None)

    def set(self, context: Context, value: ValueT) -> None:
        context._set_attribute(self, None, value)

    def delete(self, context: Context) -> None:
        context._delete_attribute(cast(Attribute[object], self), None)

    def lock(self) -> AttributeLock:
        return AttributeLock(self)


@dataclass(frozen=True)
class AttributeMap(Generic[ValueT]):
    name: str
    value_type: type[ValueT]
    index: AttributeIndex | None = None

    def __post_init__(self) -> None:
        require_name(self.name)

    def get(self, context: Context, instance: str) -> ValueT:
        return context._get_attribute(self, instance)

    def set(self, context: Context, instance: str, value: ValueT) -> None:
        context._set_attribute(self, instance, value)

    def delete(self, context: Context, instance: str) -> None:
        context._delete_attribute(cast(AttributeMap[object], self), instance)

    def lock(self, instance: str) -> AttributeLock:
        require_name(instance)
        return AttributeLock(self, instance)


@dataclass(frozen=True)
class AttributeLock:
    attribute: Attribute[Any] | AttributeMap[Any]
    instance: str | None = None


@dataclass(frozen=True)
class Condition:
    condition_id: str | None = None

    def __post_init__(self) -> None:
        validate_condition_id(self.condition_id)


@dataclass(frozen=True)
class TimerCondition(Condition):
    duration: timedelta = timedelta(0)

    def __post_init__(self) -> None:
        super().__post_init__()
        if self.duration < timedelta(0):
            raise ValueError("timer duration must not be negative")


@dataclass(frozen=True)
class ChannelCondition(Condition, Generic[ValueT]):
    channel: Channel[ValueT] | ChannelMap[ValueT] | None = None
    instance: str | None = None
    at_least: int | None = None
    at_most: int | None = None

    def __post_init__(self) -> None:
        super().__post_init__()
        if self.channel is None:
            raise ValueError("channel condition requires a channel")
        if self.at_least is None and self.at_most is None:
            raise ValueError("channel condition requires a bound")
        if self.at_least is not None and self.at_least < 0:
            raise ValueError("at_least must not be negative")
        if self.at_most is not None and self.at_most < 0:
            raise ValueError("at_most must not be negative")
        if (
            self.at_least is not None
            and self.at_most is not None
            and self.at_most < self.at_least
        ):
            raise ValueError("at_most must not be below at_least")


@dataclass(frozen=True)
class ConditionCombination:
    conditions: tuple[Condition, ...]

    @staticmethod
    def of(*conditions: Condition) -> ConditionCombination:
        return ConditionCombination(conditions)


class Timer:
    @staticmethod
    def by_duration(
        duration: timedelta,
        *,
        condition_id: str | None = None,
    ) -> Condition:
        return TimerCondition(condition_id=condition_id, duration=duration)


@dataclass(frozen=True)
class Channel(Generic[ValueT]):
    name: str
    value_type: type[ValueT]

    def __post_init__(self) -> None:
        require_name(self.name)

    def publish(self, context: Context, value: ValueT) -> None:
        context._publish_channel(self, None, value)

    def size(self, context: Context) -> int:
        return context._channel_size(cast(Channel[object], self), None)

    def results(self, context: Context) -> Sequence[ValueT]:
        return context._channel_results(self, None)

    def for_one(self, *, condition_id: str | None = None) -> Condition:
        return self.for_range(at_least=1, at_most=1, condition_id=condition_id)

    def for_n(self, count: int, *, condition_id: str | None = None) -> Condition:
        return self.for_range(
            at_least=count,
            at_most=count,
            condition_id=condition_id,
        )

    def at_least(
        self,
        count: int,
        *,
        condition_id: str | None = None,
    ) -> Condition:
        return self.for_range(at_least=count, condition_id=condition_id)

    def at_most(
        self,
        count: int,
        *,
        condition_id: str | None = None,
    ) -> Condition:
        return self.for_range(at_most=count, condition_id=condition_id)

    def for_range(
        self,
        *,
        at_least: int | None = None,
        at_most: int | None = None,
        condition_id: str | None = None,
    ) -> Condition:
        return ChannelCondition(
            condition_id=condition_id,
            channel=self,
            at_least=at_least,
            at_most=at_most,
        )


@dataclass(frozen=True)
class ChannelMap(Generic[ValueT]):
    name: str
    value_type: type[ValueT]

    def __post_init__(self) -> None:
        require_name(self.name)

    def publish(self, context: Context, instance: str, value: ValueT) -> None:
        context._publish_channel(self, instance, value)

    def size(self, context: Context, instance: str) -> int:
        return context._channel_size(cast(ChannelMap[object], self), instance)

    def results(self, context: Context, instance: str) -> Sequence[ValueT]:
        return context._channel_results(self, instance)

    def for_one(
        self,
        instance: str,
        *,
        condition_id: str | None = None,
    ) -> Condition:
        return self.for_range(
            instance,
            at_least=1,
            at_most=1,
            condition_id=condition_id,
        )

    def for_n(
        self,
        instance: str,
        count: int,
        *,
        condition_id: str | None = None,
    ) -> Condition:
        return self.for_range(
            instance,
            at_least=count,
            at_most=count,
            condition_id=condition_id,
        )

    def at_least(
        self,
        instance: str,
        count: int,
        *,
        condition_id: str | None = None,
    ) -> Condition:
        return self.for_range(
            instance,
            at_least=count,
            condition_id=condition_id,
        )

    def at_most(
        self,
        instance: str,
        count: int,
        *,
        condition_id: str | None = None,
    ) -> Condition:
        return self.for_range(
            instance,
            at_most=count,
            condition_id=condition_id,
        )

    def for_range(
        self,
        instance: str,
        *,
        at_least: int | None = None,
        at_most: int | None = None,
        condition_id: str | None = None,
    ) -> Condition:
        return ChannelCondition(
            condition_id=condition_id,
            channel=self,
            instance=instance,
            at_least=at_least,
            at_most=at_most,
        )


class WaitKind(Enum):
    SKIP_IMMEDIATELY = "skip_immediately"
    ALL_OF = "all_of"
    ANY_OF = "any_of"
    ANY_COMBINATION_OF = "any_combination_of"


@dataclass(frozen=True)
class Wait:
    kind: WaitKind
    conditions: tuple[Condition, ...] = ()
    combinations: tuple[ConditionCombination, ...] = ()

    @staticmethod
    def skip_immediately() -> Wait:
        return Wait(WaitKind.SKIP_IMMEDIATELY)

    @staticmethod
    def all_of(*conditions: Condition) -> Wait:
        return Wait(WaitKind.ALL_OF, conditions)

    @staticmethod
    def any_of(*conditions: Condition) -> Wait:
        return Wait(WaitKind.ANY_OF, conditions)

    @staticmethod
    def any_combination_of(*combinations: ConditionCombination) -> Wait:
        return Wait(
            WaitKind.ANY_COMBINATION_OF,
            combinations=combinations,
        )

File: sdk-rust/crates/dex-sdk/src/state.rs
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

use std::marker::PhantomData;
use std::time::SystemTime;

use crate::{Condition, HandlerResult, Value};

pub struct Attribute<T> {
    name: String,
    index: Option<AttributeIndex>,
    marker: PhantomData<fn() -> T>,
}

impl<T> Attribute<T> {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            index: None,
            marker: PhantomData,
        }
    }

    pub fn indexed(mut self, index: AttributeIndex) -> Self {
        self.index = Some(index);
        self
    }

    pub fn get(&self, context: &Context) -> HandlerResult<Option<T>>
    where
        T: Value,
    {
        context.get_attribute(self)
    }

    pub fn get_required(&self, context: &Context) -> HandlerResult<T>
    where
        T: Value,
    {
        self.get(context)?
            .ok_or_else(|| crate::HandlerError::new(format!("attribute {} is missing", self.name)))
    }

    pub fn set(&self, context: &mut Context, value: T) -> HandlerResult<()>
    where
        T: Value,
    {
        context.set_attribute(self, value)
    }

    pub fn delete(&self, context: &mut Context) -> HandlerResult<()> {
        context.delete_attribute(self)
    }

    pub fn lock(&self) -> AttributeLock {
        AttributeLock {
            attribute: self.name.clone(),
            instance: None,
        }
    }
}

impl<T> Clone for Attribute<T> {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            index: self.index.clone(),
            marker: PhantomData,
        }
    }
}

pub struct AttributeMap<T> {
    name: String,
    index: Option<AttributeIndex>,
    marker: PhantomData<fn() -> T>,
}

impl<T> AttributeMap<T> {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            index: None,
            marker: PhantomData,
        }
    }

    pub fn indexed(mut self, index: AttributeIndex) -> Self {
        self.index = Some(index);
        self
    }

    pub fn get(&self, context: &Context, instance: &str) -> HandlerResult<Option<T>>
    where
        T: Value,
    {
        context.get_attribute_map(self, instance)
    }

    pub fn get_required(&self, context: &Context, instance: &str) -> HandlerResult<T>
    where
        T: Value,
    {
        self.get(context, instance)?.ok_or_else(|| {
            crate::HandlerError::new(format!("attribute {}[{instance}] is missing", self.name))
        })
    }

    pub fn set(&self, context: &mut Context, instance: &str, value: T) -> HandlerResult<()>
    where
        T: Value,
    {
        context.set_attribute_map(self, instance, value)
    }

    pub fn delete(&self, context: &mut Context, instance: &str) -> HandlerResult<()> {
        context.delete_attribute_map(self, instance)
    }

    pub fn lock(&self, instance: impl Into<String>) -> AttributeLock {
        AttributeLock {
            attribute: self.name.clone(),
            instance: Some(instance.into()),
        }
    }
}

impl<T> Clone for AttributeMap<T> {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            index: self.index.clone(),
            marker: PhantomData,
        }
    }
}

pub struct Channel<T> {
    name: String,
    marker: PhantomData<fn() -> T>,
}

impl<T> Channel<T> {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            marker: PhantomData,
        }
    }

    pub fn publish(&self, context: &mut Context, value: T) -> HandlerResult<()>
    where
        T: Value,
    {
        context.publish(self, value)
    }

    pub fn size(&self, context: &Context) -> HandlerResult<usize> {
        context.channel_size(self)
    }

    pub fn condition_results(&self, context: &Context) -> HandlerResult<Vec<T>>
    where
        T: Value,
    {
        context.channel_results(self)
    }

    pub fn for_one(&self) -> Condition {
        self.range(Some(1), Some(1))
    }

    pub fn for_n(&self, count: usize) -> Condition {
        self.range(Some(count), Some(count))
    }

    pub fn at_least(&self, count: usize) -> Condition {
        self.range(Some(count), None)
    }

    pub fn at_most(&self, count: usize) -> Condition {
        self.range(None, Some(count))
    }

    pub fn range(&self, at_least: Option<usize>, at_most: Option<usize>) -> Condition {
        Condition::channel(self.name.clone(), None, at_least, at_most)
    }

    pub fn when_empty(&self) -> ChannelGuard {
        ChannelGuard {
            name: self.name.clone(),
            instance: None,
        }
    }
}

impl<T> Clone for Channel<T> {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            marker: PhantomData,
        }
    }
}

pub struct ChannelMap<T> {
    name: String,
    marker: PhantomData<fn() -> T>,
}

impl<T> ChannelMap<T> {
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            marker: PhantomData,
        }
    }

    pub fn publish(&self, context: &mut Context, instance: &str, value: T) -> HandlerResult<()>
    where
        T: Value,
    {
        context.publish_map(self, instance, value)
    }

    pub fn size(&self, context: &Context, instance: &str) -> HandlerResult<usize> {
        context.channel_map_size(self, instance)
    }

    pub fn condition_results(&self, context: &Context, instance: &str) -> HandlerResult<Vec<T>>
    where
        T: Value,
    {
        context.channel_map_results(self, instance)
    }

    pub fn for_one(&self, instance: &str) -> Condition {
        self.range(instance, Some(1), Some(1))
    }

    pub fn for_n(&self, instance: &str, count: usize) -> Condition {
        self.range(instance, Some(count), Some(count))
    }

    pub fn at_least(&self, instance: &str, count: usize) -> Condition {
        self.range(instance, Some(count), None)
    }

    pub fn at_most(&self, instance: &str, count: usize) -> Condition {
        self.range(instance, None, Some(count))
    }

    pub fn range(
        &self,
        instance: &str,
        at_least: Option<usize>,
        at_most: Option<usize>,
    ) -> Condition {
        Condition::channel(
            self.name.clone(),
            Some(instance.to_string()),
            at_least,
            at_most,
        )
    }

    pub fn when_empty(&self, instance: &str) -> ChannelGuard {
        ChannelGuard {
            name: self.name.clone(),
            instance: Some(instance.to_string()),
        }
    }
}

impl<T> Clone for ChannelMap<T> {
    fn clone(&self) -> Self {
        Self {
            name: self.name.clone(),
            marker: PhantomData,
        }
    }
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AttributeIndex {
    kind: AttributeIndexKind,
    key: Option<String>,
}

impl AttributeIndex {
    fn new(kind: AttributeIndexKind) -> Self {
        Self { kind, key: None }
    }

    pub fn keyword() -> Self {
        Self::new(AttributeIndexKind::Keyword)
    }

    pub fn full_text() -> Self {
        Self::new(AttributeIndexKind::FullText)
    }

    pub fn keyword_array() -> Self {
        Self::new(AttributeIndexKind::KeywordArray)
    }

    pub fn int() -> Self {
        Self::new(AttributeIndexKind::Int)
    }

    pub fn double() -> Self {
        Self::new(AttributeIndexKind::Double)
    }

    pub fn bool() -> Self {
        Self::new(AttributeIndexKind::Bool)
    }

    pub fn date_time() -> Self {
        Self::new(AttributeIndexKind::DateTime)
    }

    pub fn with_key(mut self, key: impl Into<String>) -> Self {
        self.key = Some(key.into());
        self
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
enum AttributeIndexKind {
    Keyword,
    FullText,
    KeywordArray,
    Int,
    Double,
    Bool,
    DateTime,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct AttributeLock {
    attribute: String,
    instance: Option<String>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ChannelGuard {
    name: String,
    instance: Option<String>,
}

#[derive(Clone, Debug, Default, Eq, PartialEq)]
pub struct PersistenceSchema {
    definitions: Vec<PersistenceDefinition>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
enum PersistenceDefinition {
    Attribute,
    AttributeMap,
    Channel,
    ChannelMap,
}

impl PersistenceSchema {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn attribute<T>(mut self, _attribute: &Attribute<T>) -> Self {
        self.definitions.push(PersistenceDefinition::Attribute);
        self
    }

    pub fn attribute_map<T>(mut self, _attribute: &AttributeMap<T>) -> Self {
        self.definitions.push(PersistenceDefinition::AttributeMap);
        self
    }

    pub fn channel<T>(mut self, _channel: &Channel<T>) -> Self {
        self.definitions.push(PersistenceDefinition::Channel);
        self
    }

    pub fn channel_map<T>(mut self, _channel: &ChannelMap<T>) -> Self {
        self.definitions.push(PersistenceDefinition::ChannelMap);
        self
    }
}

pub struct Context {
    _private: (),
}

impl Context {
    pub fn flow_id(&self) -> &str {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn run_id(&self) -> &str {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn flow_started_at(&self) -> SystemTime {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn attempt(&self) -> u32 {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn has_timer_fired(&self, _index: usize) -> bool {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn wait_for_method_failed(&self) -> bool {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn set_step_execution_local<T: Value>(
        &mut self,
        _key: &str,
        _value: T,
    ) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn step_execution_local<T: Value>(&self, _key: &str) -> HandlerResult<T> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn record_event<T: Value>(&mut self, _name: &str, _value: T) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn get_attribute<T: Value>(&self, _attribute: &Attribute<T>) -> HandlerResult<Option<T>> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn get_attribute_map<T: Value>(
        &self,
        _attribute: &AttributeMap<T>,
        _instance: &str,
    ) -> HandlerResult<Option<T>> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn set_attribute<T: Value>(
        &mut self,
        _attribute: &Attribute<T>,
        _value: T,
    ) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn set_attribute_map<T: Value>(
        &mut self,
        _attribute: &AttributeMap<T>,
        _instance: &str,
        _value: T,
    ) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn delete_attribute<T>(&mut self, _attribute: &Attribute<T>) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn delete_attribute_map<T>(
        &mut self,
        _attribute: &AttributeMap<T>,
        _instance: &str,
    ) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn publish<T: Value>(&mut self, _channel: &Channel<T>, _value: T) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn publish_map<T: Value>(
        &mut self,
        _channel: &ChannelMap<T>,
        _instance: &str,
        _value: T,
    ) -> HandlerResult<()> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn channel_size<T>(&self, _channel: &Channel<T>) -> HandlerResult<usize> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn channel_map_size<T>(
        &self,
        _channel: &ChannelMap<T>,
        _instance: &str,
    ) -> HandlerResult<usize> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn channel_results<T: Value>(&self, _channel: &Channel<T>) -> HandlerResult<Vec<T>> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }

    pub fn channel_map_results<T: Value>(
        &self,
        _channel: &ChannelMap<T>,
        _instance: &str,
    ) -> HandlerResult<Vec<T>> {
        unimplemented!("Context belongs to the Rust Worker runtime")
    }
}
